### PR TITLE
[Order] Implementation of OrderItemQuantityModifier

### PR DESCRIPTION
**Pull request by Zales0123**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | #3524
| License       | MIT
| Doc PR        | https://github.com/Sylius/Sylius-Docs/pull/400

This PR is based on https://github.com/Sylius/Sylius/pull/3771 and correspond with @michalmarcinkowski work in https://github.com/Sylius/Sylius/pull/3786.

In this PR there is ``OrderItemQuantityModifier`` implemented, that handles item units management. The idea is, to change ``OrderItem`` quantity only by adding/removing units.

**Possible improvements/modifications**
- it will be better to remove *backordered* units first, then *sold* ones

**UPDATE**

I've applied commits from https://github.com/Sylius/Sylius/pull/3808, as they're necessary to make behat scenarios green.

**UPDATE vol.2**

Docs PR mentioned.

### CODE REVIEW

File: src/Sylius/Component/Core/Model/Order.php [hunk 1]
```diff
@@ -15,6 +15,7 @@
 use Doctrine\Common\Collections\Collection;
 use Sylius\Component\Cart\Model\Cart;
 use Sylius\Component\Channel\Model\ChannelInterface as BaseChannelInterface;
+use Sylius\Component\Inventory\Model\InventoryUnitInterface;
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Model/Order.php]: Is it used?

Comment (lchrusciel) [inline src/Sylius/Component/Core/Model/Order.php]: yes, it is. My bad.

File: src/Sylius/Component/Core/spec/OrderProcessing/OrderShipmentFactorySpec.php [hunk 2]
```diff
@@ -13,10 +13,10 @@
 
 use Doctrine\Common\Collections\ArrayCollection;
 use PhpSpec\ObjectBehavior;
-use Sylius\Component\Core\Model\InventoryUnitInterface;
 use Sylius\Component\Core\Model\OrderInterface;
 use Sylius\Component\Core\Model\ShipmentInterface;
 use Sylius\Component\Core\OrderProcessing\OrderShipmentFactoryInterface;
+use Sylius\Component\Core\Model\OrderItemUnitInterface;
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/spec/OrderProcessing/OrderShipmentFactorySpec.php]: Change to alphabetical order.

File: src/Sylius/Component/Order/Model/Adjustment.php [hunk 3]
```diff
@@ -24,12 +24,17 @@ class Adjustment implements AdjustmentInterface
     /**
      * @var OrderInterface
      */
-    protected $order;
+    protected $order = null;
```
Comment (lchrusciel) [inline src/Sylius/Component/Order/Model/Adjustment.php]: Doesn't it have a null value by default?

File: src/Sylius/Component/Core/spec/OrderProcessing/OrderShipmentFactorySpec.php (hunk 2 above)
Comment (Zales0123) [inline src/Sylius/Component/Core/spec/OrderProcessing/OrderShipmentFactorySpec.php]: I'll change it in base PR ;)

File: src/Sylius/Component/Order/Model/OrderItemInterface.php [hunk 4]
```diff
@@ -86,4 +87,33 @@ public function isImmutable();
...
+
+    /**
+     * @param OrderItemUnitInterface $itemUnit
+     */
+    public function removeUnit(OrderItemUnitInterface $itemUnit);
+
+    /**
+     * @param int $index
+     *
+     * @return mixed
```
Comment (lchrusciel) [inline src/Sylius/Component/Order/Model/OrderItemInterface.php]: Why do you return anything here?

File: src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * @param OrderItemInterface $orderItem
+     * @param int $targetUnitsNumber
+     */
+    private function increaseUnitsNumber(OrderItemInterface $orderItem, $targetUnitsNumber)
+    {
+        for ($i = 0; $i < $targetUnitsNumber; $i++) {
+            $unit = $this->orderItemUnitFactory->createNew();
+            $orderItem->addUnit($unit);
+        }
```
Comment (peteward) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: Wouldn't you want to setQuantity on the $orderItem at this point? same with decrease below?

I presume quantity is being kept as an attribute of OrderItem?

Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: @peteward Yes, but it will calculated from units. :) So it will be stored in property for fast retrieval but always in sync with count(units).

Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: Sure, so ``addUnit`` is going to increate ``$quantity`` property, and ``removeUnit`` - decrease. In fact it can be done even know, doesn't it?

Comment (peteward) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: :+1:  I don't see why you shouldn't do it now

File: src/Sylius/Component/Order/Model/Adjustment.php (hunk 3 above)
Comment (Zales0123) [inline src/Sylius/Component/Order/Model/Adjustment.php]: As above, changed in https://github.com/Sylius/Sylius/pull/3771

Comment (lchrusciel) [inline src/Sylius/Component/Order/Model/Adjustment.php]: :+1:

File: src/Sylius/Component/Core/spec/OrderProcessing/OrderShipmentFactorySpec.php (hunk 2 above)
Comment (lchrusciel) [inline src/Sylius/Component/Core/spec/OrderProcessing/OrderShipmentFactorySpec.php]: :+1:

File: src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    public function modify(OrderItemInterface $orderItem, $targetQuantity)
+    {
+        $itemQuantity = $orderItem->getQuantity();
+
+        if ($targetQuantity > $itemQuantity) {
+            $this->increaseUnitsNumber($orderItem, $targetQuantity - $itemQuantity);
+        } else if ($targetQuantity < $itemQuantity) {
+            $this->decreaseUnitsNumber($orderItem, $itemQuantity - $targetQuantity);
+        }
+
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: Redundant blank line

File: src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * @param OrderItemInterface $orderItem
+     * @param int $increaseBy
+     */
+    private function increaseUnitsNumber(OrderItemInterface $orderItem, $increaseBy)
+    {
+        for ($i = 0; $i < $increaseBy; $i++) {
+            $unit = $this->orderItemUnitFactory->createForItem($orderItem);
+            $orderItem->addUnit($unit);
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: Not needed any more (unit is added to the order item in the constructor).

Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: :+1:

File: src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+     * {@inheritdoc}
+     */
+    public function modify(OrderItemInterface $orderItem, $targetQuantity)
+    {
+        $itemQuantity = $orderItem->getQuantity();
+
+        if ($targetQuantity > $itemQuantity) {
+            $this->increaseUnitsNumber($orderItem, $targetQuantity - $itemQuantity);
+        } else if ($targetQuantity < $itemQuantity) {
+            $this->decreaseUnitsNumber($orderItem, $itemQuantity - $targetQuantity);
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: We should validate if `$targetQuantity` is not negative, in other case we could have `$decreaseBy = 10 - (-10) = 20`.

File: src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifierInterface.php [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+/**
+ * @author Mateusz Zalewski <mateusz.zalewski@lakion.com>
+ */
+interface OrderItemQuantityModifierInterface
+{
+    /**
+     * @param OrderItemInterface $orderItem
+     * @param int $quantity
+     */
+    public function modify(OrderItemInterface $orderItem, $quantity);
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifierInterface.php]: `$quantity` => `$targetQuantity`

File: src/Sylius/Bundle/CoreBundle/spec/Modifier/OrderItemQuantityModifierSpec.php [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+    function it_adds_proper_number_of_order_item_units_to_order_item(
+        $orderItemUnitFactory,
+        OrderItemInterface $orderItem,
+        OrderItemUnitInterface $unit1,
+        OrderItemUnitInterface $unit2
+    ) {
+        $orderItem->getQuantity()->willReturn(1);
+
+        $orderItemUnitFactory->createForItem($orderItem)->willReturn($unit1, $unit2);
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/spec/Modifier/OrderItemQuantityModifierSpec.php]: After my comments there should be `->shouldBeCalledTimes(2)`.

File: src/Sylius/Bundle/CoreBundle/spec/Modifier/OrderItemQuantityModifierSpec.php [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    function it_does_nothing_if_target_quantity_is_equal_to_current($orderItemUnitFactory, OrderItemInterface $orderItem)
+    {
+        $orderItem->getQuantity()->willReturn(3);
+
+        $orderItemUnitFactory->createForItem(Argument::any())->shouldNotBeCalled();
+        $orderItem->addUnit(Argument::any())->shouldNotBeCalled();
+        $orderItem->removeUnit(Argument::any())->shouldNotBeCalled();
+    }
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/spec/Modifier/OrderItemQuantityModifierSpec.php]: We are missing specs for `$targetQuantity` = `0` and negative.

File: src/Sylius/Bundle/CoreBundle/spec/Modifier/OrderItemQuantityModifierSpec.php [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+        $this->modify($orderItem, 3);
+    }
+
+    function it_does_nothing_if_target_quantity_is_equal_to_current($orderItemUnitFactory, OrderItemInterface $orderItem)
+    {
+        $orderItem->getQuantity()->willReturn(3);
+
+        $orderItemUnitFactory->createForItem(Argument::any())->shouldNotBeCalled();
+        $orderItem->addUnit(Argument::any())->shouldNotBeCalled();
+        $orderItem->removeUnit(Argument::any())->shouldNotBeCalled();
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/spec/Modifier/OrderItemQuantityModifierSpec.php]: I don't see method call `$this->modify($orderItem, 3);` :smile:

File: src/Sylius/Bundle/CoreBundle/spec/Modifier/OrderItemQuantityModifierSpec.php [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        OrderItemInterface $orderItem,
+        OrderItemUnitInterface $unit1,
+        OrderItemUnitInterface $unit2,
+        OrderItemUnitInterface $unit3,
+        OrderItemUnitInterface $unit4
+    ) {
+        $orderItem->getQuantity()->willReturn(5);
+        $orderItem->getUnits()->willReturn($orderItemUnits);
+
+        $orderItemUnits->count()->willReturn(4);
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/spec/Modifier/OrderItemQuantityModifierSpec.php]: `getQuantity()->willReturn(5)` but `$orderItemUnits->count()->willReturn(4)`? I think we should be more consistent here :)

Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/spec/Modifier/OrderItemQuantityModifierSpec.php]: Ok, in fact sth gone bad here :smile:

File: src/Sylius/Bundle/OrderBundle/DependencyInjection/Configuration.php [hunk 14]
```diff
@@ -144,6 +147,19 @@ private function addResourcesSection(ArrayNodeDefinition $node)
...
                         ->end()
+                        ->arrayNode('order_item_unit')
+                            ->addDefaultsIfNotSet()
+                            ->children()
+                                ->arrayNode('classes')
+                                    ->addDefaultsIfNotSet()
+                                    ->children()
+                                        ->scalarNode('model')->defaultValue(OrderItemUnit::class)->cannotBeEmpty()->end()
+                                        ->scalarNode('interface')->defaultValue(OrderItemUnitInterface::class)->cannotBeEmpty()->end()
+                                        ->scalarNode('factory')->defaultValue(OrderItemUnitFactory::class)->end()
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/OrderBundle/DependencyInjection/Configuration.php]: I suppose we should have `->scalarNode('repository')->cannotBeEmpty()->end()` also.

File: src/Sylius/Bundle/OrderBundle/spec/Factory/OrderItemUnitFactorySpec.php [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    function it_implements_factory_interface()
+    {
+        $this->shouldImplement(OrderItemUnitFactoryInterface::class);
+    }
+
+    function it_throws_exception_while_trying_create_order_item_unit_without_order_item()
+    {
+        $this->shouldThrow('Sylius\Component\Resource\Exception\UnsupportedMethodException')->during('createNew');
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/OrderBundle/spec/Factory/OrderItemUnitFactorySpec.php]: `UnsupportedMethodException::class`

File: src/Sylius/Component/Core/Model/Order.php [hunk 16]
```diff
@@ -407,16 +363,12 @@ public function getPromotionCoupons()
     /**
      * {@inheritdoc}
      */
-    public function addPromotionCoupon($coupon)
+    public function addPromotionCoupon(BaseCouponInterface $coupon = null)
     {
         if (null === $coupon) {
             return $this;
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/Order.php]: `return;`

File: src/Sylius/Component/Core/Model/Order.php [hunk 17]
```diff
@@ -407,16 +363,12 @@ public function getPromotionCoupons()
     /**
      * {@inheritdoc}
      */
-    public function addPromotionCoupon($coupon)
+    public function addPromotionCoupon(BaseCouponInterface $coupon = null)
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/Order.php]: I think we shouldn't allow for `null`. It doesn't make sense to add `null` :)

File: src/Sylius/Component/Core/Model/Order.php [hunk 18]
```diff
@@ -427,16 +379,12 @@ public function addPromotionCoupon($coupon)
     /**
      * {@inheritdoc}
      */
-    public function removePromotionCoupon($coupon)
+    public function removePromotionCoupon(BaseCouponInterface $coupon = null)
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/Order.php]: Same as we `add`.

File: src/Sylius/Component/Core/Model/OrderInterface.php [hunk 19]
```diff
@@ -173,71 +135,51 @@ public function getCurrency();
...
-    public function addPromotionCoupon($coupon);
+    public function addPromotionCoupon(BaseCouponInterface $coupon = null);
 
     /**
-     * Removes promotion coupon.
-     *
      * @param BaseCouponInterface $coupon
      */
-    public function removePromotionCoupon($coupon);
+    public function removePromotionCoupon(BaseCouponInterface $coupon = null);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/OrderInterface.php]: Do not allow `null` here.

File: src/Sylius/Component/Core/Model/OrderItemUnit.php [hunk 20]
```diff
@@ -0,0 +1,200 @@
...
+
+    /**
+     * @var \DateTime
+     */
+    protected $createdAt;
+
+    /**
+     * @var \DateTime
+     */
+    protected $updatedAt;
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/OrderItemUnit.php]: Is there any reason to put this here instead of order component?

File: src/Sylius/Component/Core/Model/OrderItemUnit.php [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+    {
+        return $this->orderItem->getVariant();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setShippable(ShippableInterface $shippable)
+    {
+        $this->orderItem->setVariant($shippable);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/OrderItemUnit.php]: I don't think this method is necessary nor safe to use because the variant shouldn't change on order item. @pjedrzejewski do we use it at all?

File: src/Sylius/Component/Core/Model/OrderItemUnit.php [hunk 22]
```diff
@@ -0,0 +1,200 @@
...
+     */
+    public function setStockable(StockableInterface $stockable)
+    {
+        $this->orderItem->setVariant($stockable);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSku()
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/OrderItemUnit.php]: Do we need this method?

/cc @pjedrzejewski

File: src/Sylius/Component/Core/Model/OrderItemUnit.php (hunk 21 above)
Comment (Zales0123) [inline src/Sylius/Component/Core/Model/OrderItemUnit.php]: I agree with that argument :+1: @pjedrzejewski wdyt?

File: src/Sylius/Component/Core/Model/OrderItemUnit.php [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+    {
+        return $this->orderItem->getVariant();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setStockable(StockableInterface $stockable)
+    {
+        $this->orderItem->setVariant($stockable);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/OrderItemUnit.php]: Same case as here https://github.com/Sylius/Sylius/pull/3795/files#r49055673.

File: src/Sylius/Component/Core/OrderProcessing/InventoryHandler.php [hunk 24]
```diff
@@ -132,12 +128,22 @@ public function updateInventory(OrderInterface $order)
...
+            throw new \InvalidArgumentException('Quantity of units must be greater than 0.');
+        }
 
-        foreach ($units as $unit) {
-            $item->addInventoryUnit($unit);
+        for ($i = 0; $i < $quantity; $i++) {
+            $unit = $this->orderItemUnitFactory->createForItem($item);
+            $unit->setInventoryState($state);
+
+            $item->addUnit($unit);
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/OrderProcessing/InventoryHandler.php]: This is no longer needed, we add unit to the item in the constructor.

File: src/Sylius/Component/Core/spec/Model/OrderSpec.php [hunk 25]
```diff
@@ -252,11 +274,75 @@ function it_not_a_backorder_if_contains_no_backordered_units(
...
         $this->addItem($item);
 
         $this->shouldNotBeBackorder();
     }
+
+    function it_adds_and_removes_payments(PaymentInterface $payment)
+    {
+        $payment->getState()->willReturn(PaymentInterface::STATE_PENDING);
+        $payment->setOrder($this)->shouldBeCalled();
+        $payment->setOrder(null)->shouldBeCalled();
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/spec/Model/OrderSpec.php]: This should be moved after `$this->addPayment($payment);`

File: src/Sylius/Component/Core/spec/Model/OrderSpec.php [hunk 26]
```diff
@@ -252,11 +274,75 @@ function it_not_a_backorder_if_contains_no_backordered_units(
...
+        $this->addPayment($payment1);
+        $this->addPayment($payment2);
+
+        $this->getLastPayment()->shouldReturn($payment2);
+    }
+
+    function it_adds_and_removes_shipments(ShipmentInterface $shipment)
+    {
+        $shipment->setOrder($this)->shouldBeCalled();
+        $shipment->setOrder(null)->shouldBeCalled();
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/spec/Model/OrderSpec.php]: This should be moved after `$this->addShipment($shipment);`

File: src/Sylius/Component/Core/OrderProcessing/InventoryHandler.php [hunk 27]
```diff
@@ -67,17 +63,17 @@ public function __construct(
...
-                $item->removeInventoryUnit($unit);
+            foreach ($item->getUnits()->slice(0, $nbUnits - $item->getQuantity()) as $unit) {
+                $item->removeUnit($unit);
             }
         }
 
-        foreach ($item->getInventoryUnits() as $unit) {
+        foreach ($item->getUnits() as $unit) {
             if ($unit->getStockable() !== $item->getVariant()) {
                 $unit->setStockable($item->getVariant());
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/OrderProcessing/InventoryHandler.php]: This is no longer needed, as `$unit->getStockable()` returns `$this->item->getVariant()`.

File: src/Sylius/Component/Core/OrderProcessing/InventoryHandler.php [hunk 28]
```diff
@@ -67,17 +63,17 @@ public function __construct(
      */
     public function processInventoryUnits(OrderItemInterface $item)
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/OrderProcessing/InventoryHandler.php]: This method should be replaced by `OrderItemQuantityModifier`.

File: src/Sylius/Component/Order/Model/OrderItemUnit.php [hunk 29]
```diff
@@ -0,0 +1,207 @@
...
+            return $type === $adjustment->getType();
+        });
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function addAdjustment(AdjustmentInterface $adjustment)
+    {
+        if (!$this->hasAdjustment($adjustment)) {
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Order/Model/OrderItemUnit.php]:
```php
if ($this->hasAdjustment($adjustment)) {
    return;
}

$adjustment->setAdjustable($this);
...
```

File: src/Sylius/Component/Order/Model/OrderItemUnit.php [hunk 30]
```diff
@@ -0,0 +1,207 @@
...
+            $this->orderItem->recalculateUnitsTotal();
+        }
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function removeAdjustment(AdjustmentInterface $adjustment)
+    {
+        if (!$adjustment->isLocked() && $this->hasAdjustment($adjustment)) {
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Order/Model/OrderItemUnit.php]:
```php
if ($adjustment->isLocked() || !$this->hasAdjustment($adjustment)) {
    return;
}

$adjustment->setAdjustable(null);
...
```

File: src/Sylius/Component/Order/spec/Model/OrderItemSpec.php [hunk 31]
```diff
@@ -26,14 +30,14 @@ function it_is_initializable()
         $this->shouldHaveType('Sylius\Component\Order\Model\OrderItem');
     }
 
-    function it_implements_Sylius_order_item_interface()
+    function it_implements_sylius_order_item_interface()
```
Comment (michalmarcinkowski) [inline src/Sylius/Component/Order/spec/Model/OrderItemSpec.php]: Should be reverted. We unofficialy set up convention to write proper names from capital letter,

File: src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php [hunk 32]
```diff
@@ -0,0 +1,76 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function modify(OrderItemInterface $orderItem, $targetQuantity)
+    {
+        $itemQuantity = $orderItem->getQuantity();
+
+        if ($targetQuantity > $itemQuantity) {
+            $this->increaseUnitsNumber($orderItem, $targetQuantity - $itemQuantity);
+        } else if ($targetQuantity < $itemQuantity) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: Remove else. And add early return.

Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: We've agreed with @michalmarcinkowski that ``else if`` statement is acceptable here. Instead, we would have to add two ``if``'s with two early return and I don't think it will be more readable than this ^^. I hate ``else`` as much as you, but here I think it's lesser evil...

File: src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php [hunk 33]
```diff
@@ -0,0 +1,76 @@
...
+        }
+    }
+
+    /**
+     * @param OrderItemInterface $orderItem
+     * @param int $decreaseBy
+     */
+    private function decreaseUnitsNumber(OrderItemInterface $orderItem, $decreaseBy)
+    {
+        foreach ($orderItem->getUnits() as $unit) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: Why didn't you use for loop here?

Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: See this comment: https://github.com/Sylius/Sylius/pull/3808#commitcomment-15241055

File: src/Sylius/Component/Core/OrderProcessing/InventoryHandler.php (hunk 27 above)
Comment (Zales0123) [inline src/Sylius/Component/Core/OrderProcessing/InventoryHandler.php]: I don't know how on earth this is here, I'm sure I've deleted this before... Probably some inaccuracy while cherry-picking.

File: src/Sylius/Bundle/OrderBundle/DependencyInjection/Configuration.php (hunk 14 above)
Comment (Zales0123) [inline src/Sylius/Bundle/OrderBundle/DependencyInjection/Configuration.php]: Done :+1:

File: src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php (hunk 32 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: @lchrusciel it isn't `else` it's `else if`, so different behavior :smile: I don't think `else if` is evil :)

Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Modifier/OrderItemQuantityModifier.php]: Well, ``else`` is part of ``else if``, so I think it's lesser evil but still evil :smile: :smiling_imp: However acceptable in this case

File: src/Sylius/Component/Core/Model/OrderItemUnit.php (hunk 20 above)
Comment (Zales0123) [inline src/Sylius/Component/Core/Model/OrderItemUnit.php]: I've discussed it with @pjedrzejewski and we think that for now there is no need to make ``OrderItemUnit`` timestampable in *Order* component, however, it have to has these fields in *Core*, as ``OrderItemUnitInterface`` extends ``InventoryUnitInterface``.

File: src/Sylius/Component/Core/Model/OrderItemUnit.php (hunk 22 above)
Comment (Zales0123) [inline src/Sylius/Component/Core/Model/OrderItemUnit.php]: All these unnecessary methods and service will be cleared in separate PR, I suppose ;)

File: src/Sylius/Component/Core/Model/Order.php (hunk 17 above)
Comment (Zales0123) [inline src/Sylius/Component/Core/Model/Order.php]: Unfortunately, we must allow ``null`` value :/ After adding missing typehint, it throws exception if null is passed to function and it somehow happens when we type wrong coupon on cart summary page. It should be obviously changed, but not in this PR.

Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/Order.php]: :+1:

File: src/Sylius/Component/Core/Model/OrderItemUnit.php (hunk 20 above)
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/OrderItemUnit.php]: :+1:

File: src/Sylius/Component/Core/Model/OrderItemUnit.php (hunk 22 above)
Comment (michalmarcinkowski) [inline src/Sylius/Component/Core/Model/OrderItemUnit.php]: :+1:

File: src/Sylius/Bundle/CartBundle/spec/EventListener/CartSubscriberSpec.php [hunk 34]
```diff
@@ -37,15 +40,59 @@ function it_is_initializable()
         $this->shouldHaveType('Sylius\Bundle\CartBundle\EventListener\CartSubscriber');
     }
 
-    function it_should_add_a_item_to_a_cart_from_event(CartItemEvent $event, CartInterface $cart, CartItemInterface $cartItem)
-    {
+    function it_adds_a_item_to_a_cart_from_event_if_not_already_exist(
```
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CartBundle/spec/EventListener/CartSubscriberSpec.php]: adds_a**n**_

**Comment by michalmarcinkowski** [comment]
General remarks:

* In `clearItems` after `clear` we should recalculate items total on **Order** model
* `setOrder` method on **OrderItem** should ensure bidirectional relation see similar changes (https://github.com/michalmarcinkowski/Sylius/commit/e19b442c2f4c90bd944665545ac46b3a6d437578)

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml [hunk 35]
```diff
@@ -410,6 +410,7 @@
         <service id="sylius.promotion_action.add_product" class="%sylius.promotion_action.add_product.class%">
             <argument type="service" id="sylius.factory.order_item" />
             <argument type="service" id="sylius.repository.product_variant" />
+            <argument type="service" id="sylius.modifier.order_item_quantity" />
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml]: I think we should name it just ``sylius.order_item_quantity_modifier``. Why? Because we should only group services, which really do the same thing. It is true for factory, repository, manager, controller etc. But Modifier in Sylius terms is very vague. We don't have any other modifiers and I doubt we will have anytime soon and if yes, then there is a low chance that they will be doing similar thing to this one. :) WDYT? /cc @michalmarcinkowski

Comment (Zales0123) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml]: :+1: I totally agree with this argumentation.

File: src/Sylius/Bundle/OrderBundle/DependencyInjection/SyliusOrderExtension.php [hunk 36]
```diff
@@ -41,5 +39,10 @@ public function load(array $config, ContainerBuilder $container)
         foreach ($configFiles as $configFile) {
             $loader->load($configFile);
         }
+
+        $orderItemQuantityDataMapper = $container->getDefinition('sylius.data_mapper.order_item_quantity');
+        $orderItemType = $container->getDefinition('sylius.form.type.order_item');
+
+        $orderItemType->addArgument($orderItemQuantityDataMapper);
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/OrderBundle/DependencyInjection/SyliusOrderExtension.php]: We should add a ``Reference`` here, not the definition.

File: src/Sylius/Bundle/OrderBundle/Form/DataMapper/OrderItemQuantityDataMapper.php [hunk 37]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        $this->propertyPathDataMapper->mapDataToForms($data, $forms);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function mapFormsToData($forms, &$data)
+    {
+        $nonQuantityForms = array();
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/OrderBundle/Form/DataMapper/OrderItemQuantityDataMapper.php]: $formsOtherThanQuantity?

File: src/Sylius/Bundle/OrderBundle/DependencyInjection/SyliusOrderExtension.php (hunk 36 above)
Comment (Zales0123) [inline src/Sylius/Bundle/OrderBundle/DependencyInjection/SyliusOrderExtension.php]: It's already fixed in related PR https://github.com/Sylius/Sylius/pull/3842#discussion-diff-49657641.  Shall I fix this here too?

File: src/Sylius/Bundle/OrderBundle/Modifier/OrderItemQuantityModifier.php [hunk 38]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        $this->orderItemUnitFactory = $orderItemUnitFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function modify(OrderItemInterface $orderItem, $targetQuantity)
+    {
+        $itemQuantity = $orderItem->getQuantity();
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/OrderBundle/Modifier/OrderItemQuantityModifier.php]: $currentQuantity?

File: src/Sylius/Bundle/OrderBundle/Resources/config/services.xml [hunk 39]
```diff
@@ -23,6 +23,10 @@
         <parameter key="sylius.callback.complete_order.class">Sylius\Bundle\OrderBundle\StateMachineCallback\CompleteOrderCallback</parameter>
 
         <parameter key="sylius.originator.class">Sylius\Component\Originator\Originator\Originator</parameter>
+
+        <parameter key="sylius.modifier.order_item_quantity.class">Sylius\Bundle\OrderBundle\Modifier\OrderItemQuantityModifier</parameter>
+        <parameter key="sylius.data_mapper.property_path.class">Symfony\Component\Form\Extension\Core\DataMapper\PropertyPathMapper</parameter>
+        <parameter key="sylius.data_mapper.order_item_quantity.class">Sylius\Bundle\OrderBundle\Form\DataMapper\OrderItemQuantityDataMapper</parameter>
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/OrderBundle/Resources/config/services.xml]: ``sylius.form.data_mapper<xyz>``

File: src/Sylius/Bundle/CoreBundle/Resources/config/services.xml (hunk 35 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/CoreBundle/Resources/config/services.xml]: :+1:

File: src/Sylius/Bundle/OrderBundle/Modifier/OrderItemQuantityModifier.php (hunk 38 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/OrderBundle/Modifier/OrderItemQuantityModifier.php]: :+1:

File: src/Sylius/Bundle/OrderBundle/Resources/config/services.xml (hunk 39 above)
Comment (michalmarcinkowski) [inline src/Sylius/Bundle/OrderBundle/Resources/config/services.xml]: :+1:

File: src/Sylius/Bundle/OrderBundle/DependencyInjection/SyliusOrderExtension.php (hunk 36 above)
Comment (pjedrzejewski) [inline src/Sylius/Bundle/OrderBundle/DependencyInjection/SyliusOrderExtension.php]: Ok, let's leave it.

**Comment by pjedrzejewski** [comment]
Thank you Mateusz! :+1: We are smashing this thing! ;)
